# [CptS 215 Data Analytics Systems and Algorithms](https://piazza.com/wsu/fall2017/cpts215/home)
[Washington State University](https://wsu.edu)

[Gina Sprint](http://eecs.wsu.edu/~gsprint/)
## MA4 Queue Analysis (50 pts)
<mark>Due:</mark>

### Learner Objectives
At the conclusion of this micro assignment, participants should be able to:
* Analyze stack and queue data structures for efficiency
* Implement a queue using two stacks
* Compare/contrast different implementations of a queue ADT

### Prerequisites
Before starting this micro assignment, participants should be able to:
* Write object oriented Python code
* Write Markdown and code cells in Jupyter Notebook
* Understand the stack ADT and implement a stack
* Understand the queue ADT and implement a queue

### Acknowledgments
Content used in this assignment is based upon information in the following sources:
* Carl Kingsford's [Basic Data Structures](https://www.engage-csedu.org/find-resources/problem-set-1-basic-data-structures) problem set.

## Overview and Requirements
For this micro assignment, you are going to download this Jupyter Notebook and answer the following questions. Your answer for a problem should be in a cell *immediately* after the problem cell. *Do not modify the problem cell.*

We are going to explore the efficiency of two different queue implementations. This micro assignment includes conceptional questions and programming.

### Conceptual Questions
Suppose you are given a (strange) computer that can only perform the following instructions (in addition to if and while):
* `S = create_stack()` create stack makes a new stack `S`
* `i = S.pop()` removes the top item from stack `S` and places it in variable `i`
* `S.push(i)` makes item `i` the top item in stack `S`

Solve the following problems and *justify* your answers:
1. (10 pts) Show how you can use these operations to implement a queue (operations `Q = create_queue()`, `enqueue(i)`, `i = dequeue()`)
    * A picture might help to explain your answer
    * Hint: take a look at the following image:
<img src="http://www.algoqueue.com/algoqueue/members/get_uploaded_image.load/149" width="500">
(image from [http://www.algoqueue.com/algoqueue/members/get_uploaded_image.load/149](http://www.algoqueue.com/algoqueue/members/get_uploaded_image.load/149))
1. (5 pts) What's the worst case running time of your dequeue implementation?
1. (5 pts) Over a series of `n` enqueues followed by `n` dequeues, how many `pop()` operations does your implementation perform?

### Coceptual Questions, Answer:
1. We would use stack `pop()` as our queue `dequeue()` opertaion. However, using stack to perform `enqueue()` would be ineffiecient, and would require for us to go through whole stack, while saving it. To explain more, we would have to create a temporary stack that will save our data until we reach the end of stack. Code for `enqueue()`

Assuming we have `is_empty()` method that tells if stack is empty or not.

```py
def enqueue(self, thing):
    S = create_stack()
    # get all elements out of the stack, saving them.
    while is not self.is_empty():
        S.stack.push(self.stack.pop())
    self.stack.push(thing) # add our element
    
    # Put all elements back to the stack
    while is not S.is_empty():
        self.push(S.pop())
```

2. `dequeue()` will always be O(1) because it calls `pop()` on stack which is O(1) operation time always.
3. each `enqueue()` is O(N) time (Technically O(2N), but 2 doesn't count in Big O). So for `n` enqueues that is N * O(N) = $O(N^{2})$. Then we follow n of dequeue which is O(N) in total. So we have $O(N^{2}) + O(N)$ = O(N)(O(N) + 1) which is just $O(N^{2})$.
### Implementation Questions
Write a program that implements a queue using a standard list implementation (see [M&R 3.12](http://interactivepython.org/runestone/static/pythonds/BasicDS/ImplementingaQueueinPython.html)) and a queue using your solution to conceptual question #1. For the latter, you must implement a stack using a standard list implementation (see [M&R 3.5](http://interactivepython.org/runestone/static/pythonds/BasicDS/ImplementingaStackinPython.html)).

Generate a sequence of `enqueue`s and `dequeue`s to test your two queue implementations. To generate the test sequence, randomly `enqueue` and `dequeue` strings from [words.txt](https://raw.githubusercontent.com/gsprint23/cpts215/master/microassignments/files/words.txt), a file containing all 118,309 valid crossword puzzle words, one on each line. Evaluate the differences between the two implementations by performing the following:
1. Using [`timeit()`](https://docs.python.org/3/library/timeit.html), compare the running time for each queue implementation operating on your test sequence. Vary the size of your test sequence.
    * Note: Make sure you are using the same test sequence for each implementation! Also, remove all frivolous code from your implementations (e.g. `print()` statements), as these can affect the timing!
1. Write code to test your answer to conceptual question #2. Write up your observations.
1. Write code to test your answer to conceptual question #3. Write up your observations.

In [3]:
#!/usr/bin/python3
import timeit
import pandas as pd

class Stack:
    def __init__(self):
        self.list = list()
    
    def push(self, thing):
        self.list.append(thing)
    
    def pop(self):
        return self.list.pop()
    
    def is_empty(self):
        return len(self.list) == 0

    def print(self):
        print_stuff = ""
        for t in self.list:
            print_stuff += "{} ".format(t)
        print(print_stuff)

class QueueList:
    def __init__(self):
        self.list = list()
    
    def enqueue(self, thing):
        self.list.insert(0, thing)
    
    def dequeue(self):
        return self.list.pop()
    
    def print(self):
        print_stuff = ""
        for t in self.list:
            print_stuff += "{} ".format(t)
        print(print_stuff)

class QueueStack:
    def __init__(self):
        self.stack = Stack()
    
    def enqueue(self, thing):
        S = Stack()
        # get all elements out of the stack, saving them.
        while not self.is_empty():
            S.push(self.stack.pop())
        self.stack.push(thing) # add our element
        
        # Put all elements back to the stack
        while not S.is_empty():
            self.stack.push(S.pop())

    def dequeue(self):
        return self.stack.pop()
    
    def is_empty(self):
        return self.stack.is_empty()
    
    def print(self):
        print_stuff = ""
        S = Stack()

        # get all elements out of the stack, saving them.
        while not self.is_empty():
            S.push(self.stack.pop())

        # Put all elements back to the stack
        while not S.is_empty():
            element = S.pop()
            self.stack.push(element)
            print_stuff += "{} ".format(element)
        print(print_stuff)

if __name__ == "__main__":

    config = [500, 1000, 5000]
    S_queuelist = QueueList()
    S_queuestack = QueueStack()
    data = []
    for c in config:
        for f in range(1, c + 1):
            S_queuelist.enqueue(f)
            S_queuestack.enqueue(f)
        
        temp_d = ["Queue using List N={}".format(c)]
        start = timeit.default_timer()
        S_queuelist.enqueue(5)
        temp_d.append(timeit.default_timer() - start)
        start = timeit.default_timer()
        S_queuelist.dequeue()
        temp_d.append(timeit.default_timer() - start)
        data.append(temp_d)

        temp_d = ["Queue using Stack N={}".format(c)]
        start = timeit.default_timer()
        S_queuestack.enqueue(5)
        temp_d.append(timeit.default_timer() - start)
        start = timeit.default_timer()
        S_queuestack.dequeue()
        temp_d.append(timeit.default_timer() - start)
        data.append(temp_d)
    data_selected = pd.DataFrame(data, columns=["Name", "Enqueue took in Seconds", "Dequeue took in Seconds"])
    print(data_selected)

                       Name  Enqueue took in Seconds  Dequeue took in Seconds
0    Queue using List N=500                 0.000002             5.541132e-07
1   Queue using Stack N=500                 0.000426             8.311698e-07
2   Queue using List N=1000                 0.000002             5.541132e-07
3  Queue using Stack N=1000                 0.001203             5.541132e-07
4   Queue using List N=5000                 0.000002             2.770565e-07
5  Queue using Stack N=5000                 0.005302             5.541132e-07


2. for `dequeue()` it the difference is small enough not to care, because it should be O(1) in Both List and Stack implementations of Queue.
3. From Table of data above we can notice that Queue List is growing slowly from size of 500 to 5000 it did not change in its `enqueue()` at all. Yet our Queue using Stack implementation from size of 500 to 5000 changed by `0.005302 - 0.000426 = 0.004876`. So our Queue List implementation of `enqueue()` is O(1) as it did not even change. Our Queue Stack implementation of `enqueue()` is $O(N^{2})$ as we know from Conceptual question answer 3.

## Bonus (5 pts)
Perform additional analysis and comparisons of your two queue implementations above to the following:
1. Linked lists implementations of the stacks and queue (you must write the implementation yourself)
1. Python's `deque` container from `collections`

Include `timeit()` results and a write up of your observations.

In [8]:
#!/usr/bin/python3
import timeit
import collections
import pandas as pd

class Node:
    def __init__(self, data):
        self.next = None
        self.data = data

class LinkedList:
    def __init__(self):
        self.head = None
    
    def pop(self):
        if self.head is None:
            return None
        else:
            n = self.head
            while n.next.next is not None:
                n = n.next
            n.next = None
            return n
    
    def Add(self, thing):
        if self.head is None:
            self.head = thing
        else:
            thing.next = self.head
            self.head = thing
    
    def is_empty(self):
        return self.head == None

    def print(self):
        print_stuff = ""
        n = self.head
        while n.next is not None:
            print_stuff += "{} ".format(n.data)
            n = n.next
            
        print(print_stuff)

class QueueList:
    def __init__(self):
        self.list = LinkedList()
    
    def enqueue(self, thing):
        self.list.Add(Node(thing))
    
    def dequeue(self):
        return self.list.pop()
    
    def print(self):
        self.list.print()

class QueueDeque:
    def __init__(self):
        self.list = collections.deque()
    
    def enqueue(self, thing):
       self.list.appendleft(thing)

    def dequeue(self):
        return self.list.pop()

if __name__ == "__main__":

    config = [500000, 1500000, 2500000]
    S_queuelist = QueueList()
    S_queuestack = QueueDeque()
    data = []
    for c in config:
        for f in range(1, c + 1):
            S_queuelist.enqueue(f)
            S_queuestack.enqueue(f)
        
        temp_d = ["Queue using LinkedList N={}".format(c)]
        start = timeit.default_timer()
        S_queuelist.enqueue(5)
        temp_d.append(timeit.default_timer() - start)
        start = timeit.default_timer()
        S_queuelist.dequeue()
        temp_d.append(timeit.default_timer() - start)
        data.append(temp_d)

        temp_d = ["Queue using Deque N={}".format(c)]
        start = timeit.default_timer()
        S_queuestack.enqueue(5)
        temp_d.append(timeit.default_timer() - start)
        start = timeit.default_timer()
        S_queuestack.dequeue()
        temp_d.append(timeit.default_timer() - start)
        data.append(temp_d)
    data_selected = pd.DataFrame(data, columns=["Name", "Enqueue took in Seconds", "Dequeue took in Seconds"])
    print(data_selected)

                               Name  Enqueue took in Seconds  \
0   Queue using LinkedList N=500000                 0.000001   
1        Queue using Deque N=500000                 0.000001   
2  Queue using LinkedList N=1500000                 0.000002   
3       Queue using Deque N=1500000                 0.000002   
4  Queue using LinkedList N=2500000                 0.000001   
5       Queue using Deque N=2500000                 0.000002   

   Dequeue took in Seconds  
0                 0.024640  
1                 0.000001  
2                 0.098663  
3                 0.000001  
4                 0.222153  
5                 0.000001  


From the data above we can notice that I had to increase data samples to 500000, 1500000, 2500000. Because both implementations are faster than implementations before. From analyzing data both LinkedList Queue and Deque Queue `enqueue()` was O(1) as it is constant, backed up by our data.

Howver, if we look at `dequeue()` we notice that Deque Queue implementation remained the same, which means it is O(1). Yet LinkedList Queue implementation was increasing, because it takes O(N) time to walk through linked list until we reach last element. This implementation can be optimized by adding `tail` variable to the `LinkedList` class which will keep track of last Node for fast `pop()` and `append()` operations.

## Submitting Assignments
1.	Use the Blackboard tool https://learn.wsu.edu to submit your assignment. You will submit your code to the corresponding programming assignment under the "Content" tab. You must upload your solutions as `<your last name>_ma4.zip` by the due date and time.
2.	Your .zip file should contain your .ipynb file and a .html file representing your Notebook as a webpage (File->Download as->HTML). Also include [words.txt](https://raw.githubusercontent.com/gsprint23/cpts215/master/microassignments/files/words.txt) in your .zip file.

## Grading Guidelines
This assignment is worth 50 points + 5 points bonus. Your assignment will be evaluated based on a successful compilation and adherence to the program requirements. We will grade according to the following criteria:
* 20 pts for answering the conception questions
* 5 pts for correct implementation of the list-based queue
* 10 pts for correct implementation of the stack-based queue
* 5 pts for `timeit()` results
* 5 pts for observation write ups
* 5 pts for for adherence to proper programming style and comments established for the class